Adding dedicated mute button

jaikb opened the pull request:
Fixes #14 
# Changes
* I added a transparent mute button like the `reload` button in the `TwitchVideoView`.
* Added the ability to tap the mute button again (while the video is muted) for it to revert back to the previous volume the page was set at
* Button changes to mute logo if you drag the volume slider to zero as well

# Design Thoughts
I thought about just moving the volume slider straight onto the window view itself, instead of down below with the `PlayerControls`. I decided not to make any big changes unless approved by you @agg23. Let me know what you think of this implementation, and if you prefer something different I would be more than willing to implement that. Thanks!

agg23 commented:
Thanks. I'm not sure I really like the semantics of this button being overlaid on the video, but I'm not sure there's another good option. This is what I was planning on doing anyway.

### Suggested Changes

- Move the mute button over to the top left side
- If the on-window overlay button will display the muted glyph, the icon in the ornament should too
- In some scenarios a 0 volume does not correspond to a mute icon (and toggle state). In this state the button does nothing
- If you open the volume slider, press the mute button (so the video is muted), then press the mute button again, the volume slider will not reflect the volume change. The volume slider becomes completely desynced at this point, with both the button icon and the actual slider no longer showing the correct value, even after the UI is hidden

jaikb commented:
Before I implement these changes, I wanted to share a video of how `Juno` for the YouTube app does the volume slider overlaid on their video. Would you (@agg23) prefer this sort of functionality instead? 

Uploading Trimmed AVP Juno.mov…

agg23 commented:
I was avoiding doing it how Juno did because it feels not very "Apple-y" to me. What does the native video chrome or Apple TV app do? That's ideally what we should mimic.

jaikb commented:
> I was avoiding doing it how Juno did because it feels not very "Apple-y" to me. What does the native video chrome or Apple TV app do? That's ideally what we should mimic.

The Apple TV app on the AVP does it the same way Juno does it by having a volume slider on top of the video, _but_ there is no option to mute it by hitting the volume button, so just a slider no dedicated mute button. Do you (@agg23) still want to continue doing it the way I have initially suggested, or try to do it the Apple TV way but with a mute option (how Juno does it basically)?

agg23 commented:
I played with it in Apple TV. I'm really surprised they made that decision to put the volume slider on the video, but I should have followed their lead.

So, I would do it like Juno, with the overlaid volume slider, where tapping the glyph mutes and unmutes it. I can do the work if you'd like, or you can whenever you get around to it. I will hopefully be cutting a release in the next few days, assuming I can fix some of the crashes. Thanks for the help.

----

By the way, you don't need to @ me. I will automatically get notifications of issues on my repo, unless I turn it off.

jaikb commented:
Added some changes so let me know what you think

## Changes

-  Moved Reload Symbol to Top Left
- Removed the volume slider from the `PlayerControlView`
- Added a volume slider to the top right
- Made it so you can tap the volume button to mute the slider which also changes the symbol to the mute symbol. If you tap it again it will reset to the previous volume you had when you muted it

I think I left some redundant code in there that we probably don't need since there's no open and close animation or wanting to monitor that change. I'm also getting more familiar with SwiftUI, so didn't want to break something. Let me know what you think! Thanks :)

agg23 commented on Sila/View/TwitchVideoView/TwitchVideoView.swift:
```diff
@@ -46,23 +49,53 @@ struct TwitchVideoView: View {
                     }
                 }
-                .overlay(alignment: .topTrailing) {
+                .overlay(alignment: .topLeading) {
```
Use a single `.overlay(alignment: .top)` with a `HStack` and a `Spacer`.

agg23 commented on Sila/View/TwitchVideoView/TwitchVideoView.swift:
```diff
@@ -46,23 +49,53 @@ struct TwitchVideoView: View {
-                                Image(systemName: Icon.refresh)
+                        @State var isMuted = self.player.volume == 0
+                        
+                        VStack {
```
I don't think this is doing anything.

agg23 commented on Sila/View/TwitchVideoView/TwitchVideoView.swift:
```diff
@@ -46,23 +49,53 @@ struct TwitchVideoView: View {
-                                Text("Reload")
-                            } icon: {
-                                Image(systemName: Icon.refresh)
+                        @State var isMuted = self.player.volume == 0
```
This is unused, and this doesn't work. I'm surprised it even compiles.

agg23 commented on Sila/View/TwitchVideoView/TwitchVideoView.swift:
```diff
@@ -46,23 +49,53 @@ struct TwitchVideoView: View {
+                                    self.volume = 0
+                                }
+                            }
+                            .padding([.trailing, .top], 40)
```
nit: Apply styling before logic modifiers, so this would be `PopupVolumeSlider().padding().onChange()...`

agg23 commented on Sila/View/TwitchVideoView/Controls/PopupVolumeSlider.swift:
```diff
@@ -68,21 +74,11 @@ struct PopupVolumeSlider: View {
                 .opacity(self.isPresented ? 1.0 : 0.0)
                 .animation(.easeInOut(duration: 0.2), value: self.isPresented)
```
You can remove the animation, opacity, and `isPresented` stuff

agg23 commented on Sila/View/TwitchVideoView/Controls/PlayerControlsView.swift:
```diff
@@ -41,26 +41,6 @@ struct PlayerControlsView: View {
             StreamableVideoStatusControlView(player: self.player, streamableVideo: self.streamableVideo)
```
`@State private var volume: CGFloat = 0.5` is unused (this comment is on the wrong line because Github is stupid)

agg23 replied:
`@State private var volumePreventClose = false` isn't properly used, but it shouldn't be removed. This needs to be hooked into the "volume slider is active" property to prevent the UI from dismissing while you're sliding.

agg23 commented on Sila/View/TwitchVideoView/Controls/PopupVolumeSlider.swift:
```diff
@@ -9,8 +9,9 @@ import SwiftUI
-    @State private var isPresented = false
+    @State private var isPresented = true
     @State private var buttonSize = CGSize.zero
+    @State private var previousVolume: CGFloat
```
Initialize to `0.5`, so there's something available?

agg23 commented on Sila/View/TwitchVideoView/Controls/PopupVolumeSlider.swift:
```diff
@@ -9,8 +9,9 @@ import SwiftUI
 
 struct PopupVolumeSlider: View {
-    @State private var isPresented = false
+    @State private var isPresented = true
```
Remove

agg23 commented on Sila/View/TwitchVideoView/Controls/PopupVolumeSlider.swift:
```diff
@@ -20,15 +21,12 @@ struct PopupVolumeSlider: View {
-            CircleBackgroundLessButton(systemName: self.systemName, variableValue: self.volume.wrappedValue, tooltip: "Volume") {
-                self.isPresented = true
-                self.resetTimer()
+            CircleBackgroundLessButton(systemName: Icon.volume, variableValue: self.volume.wrappedValue, tooltip: "Volume") {
```
Since there is no longer an animation, you can get rid of most of this stuff. This button should be the actual mute button, and it should be in a `HStack` with the volume slider. No need for an overlay

agg23 reviewed (changes requested):
The feature generally looks good. Lots of little nitpicks to make it perfect.

Bugs:
- Mute button doesn't do anything if the video started out muted. It should jump to something like 50% volume
- Actively dragging the volume slider doesn't prevent hiding the UI
- The transition animation when pressing the mute button, between the mute and volume icon, is very slow. Possibly shouldn't have an animation at all
- When muted, when the UI fades out, the mute button icon changes from mute to the normal volume icon for some reason
- I don't think there should be a mute icon at all for that UI; just use the volume icon with the parameter at 0. This is what Apple does, and it's less jarring

Thanks for the help!

agg23 commented:
I have completed this in a separate commit. I'm very sorry to supersede you, jaikb, but I want to get a release out soon and I need the feature. You can see what I did and compare it to your work: https://github.com/agg23/Sila/commit/bd868758c166ee01b235439facbd5280043d996c

jaikb commented:
Please don't be sorry! I totally understand. I was very delayed in making the changes you requested since I was traveling to India, and also had to go back home for my sister's graduation. What's funny is I was going to finish your changes today but no worries at all. I hope my changes helped you a bit? Regardless, was wondering if any other issue you would appreciate me working on as I will start working on that now. Thanks again!

agg23 commented:
* If you wanted something fiddly to mess with, you could try working on #29.
* Fixing #21 is pretty important, but I'm not sure how to debug it.
* If you want to do something with SwiftUI, you could look into #11. Two people have requested it (which ultimately isn't that many), and I've personally never seen that functionality in Twitch, but if we had a nice way to expose it, it seems like a decent feature.
